Move MacCatalyst to runtime-extra-platforms pipeline

MaximLipnin opened the pull request:
Since https://github.com/dotnet/runtime/issues/63572 no longer happens in CI, we can promote MacCatalyst templates from the `runtime-staging` to the `runtime-extra-platforms` pipeline.

For now we will keep MacCatalyst arm64 legs in runtime-staging as they are not stable on OSX 11.00 queues (see https://github.com/dotnet/runtime/pull/64452#issuecomment-1034120637)

MaximLipnin committed:
Move MacCatalyst to runtime-extra-platforms pipeline

MaximLipnin commented:
/azp run runtime-extra-platforms

steveisok commented:
Not sure if the MacCatalyst problem is totally solved.  One of the functional tests failed under mysterious circumstances.

Commit:
Since MacCatalyst is moving out of runtime-staging, we'll need to add smoke tests to runtime.yml.  Do the same to iOS_arm64 and tvOS_arm64 as they were missing previously.

Commit:
Yml syntax

Commit:
More yml tweaks

steveisok commented:
/azp run runtime-extra-platforms

safern commented on eng/pipelines/runtime-staging.yml:
```diff
@@ -106,94 +106,6 @@ jobs:
-    buildConfig: Release
-    runtimeFlavor: mono
-    platforms:
-    - MacCatalyst_x64
```
Since the runtime-extra platforms leg doesn't run on every PR, should we add a mac catalyst leg to the main runtime pipeline or the dev-innerloop pipeline that DOES NOT RUN TESTS, just to protect the vertical build and to make sure we don't break developer inerloop scenarios?

MaximLipnin commented:
@premun Could you please look at MacCatalyst arm64 failure? It's the functional tests but they seem to fail randomly.
Some recent example: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-64452-merge-bb3ef8f27ca448e9aa/iOS.Simulator.Interpreter.Test/1/console.4367e664.log?sv=2019-07-07&se=2022-02-21T00%3A13%3A12Z&sr=c&sp=rl&sig=FOQyuT45d%2BVvzFoPbY41ZTMB1S8f%2B5y6qqSGfURaqyY%3D

```
+ sudo launchctl asuser 503 sudo -u helix-runner sh ./xharness-runner.apple.sh --target maccatalyst --command-timeout 2280 --timeout 00:30:00 --launch-timeout 00:30:00 --expected-exit-code 42 --app /tmp/helix/working/BDE009DC/w/ADB609AA/e/iOS.Simulator.Interpreter.Test.app --output-directory /tmp/helix/working/BDE009DC/w/ADB609AA/uploads
XHarness command issued: apple run --app /tmp/helix/working/BDE009DC/w/ADB609AA/e/iOS.Simulator.Interpreter.Test.app --output-directory /tmp/helix/working/BDE009DC/w/ADB609AA/uploads --target maccatalyst --timeout 00:30:00 --xcode /Applications/Xcode130.app -v --expected-exit-code 42
[16:14:49] info: Preparing run for maccatalyst
[16:14:49] info: Getting app bundle information from '/tmp/helix/working/BDE009DC/w/ADB609AA/e/iOS.Simulator.Interpreter.Test.app'
[16:14:49] dbug: 
[16:14:49] dbug: Running /usr/libexec/PlistBuddy -c "Print CFBundleName" /tmp/helix/working/BDE009DC/w/ADB609AA/e/iOS.Simulator.Interpreter.Test.app/Contents/Info.plist
[16:14:49] dbug: Process PlistBuddy exited with 0
[16:14:49] dbug: 
[16:14:49] dbug: Running /usr/libexec/PlistBuddy -c "Print CFBundleIdentifier" /tmp/helix/working/BDE009DC/w/ADB609AA/e/iOS.Simulator.Interpreter.Test.app/Contents/Info.plist
[16:14:49] dbug: Process PlistBuddy exited with 0
[16:14:49] dbug: 
[16:14:49] dbug: Running /usr/libexec/PlistBuddy -c "Print UIRequiredDeviceCapabilities" /tmp/helix/working/BDE009DC/w/ADB609AA/e/iOS.Simulator.Interpreter.Test.app/Contents/Info.plist
[16:14:49] dbug: Process PlistBuddy exited with 1
[16:14:49] dbug: Property UIRequiredDeviceCapabilities not present in Info.plist, assuming 32-bit is not supported
[16:14:49] dbug: 
[16:14:49] dbug: Running /usr/libexec/PlistBuddy -c "Print CFBundleExecutable" /tmp/helix/working/BDE009DC/w/ADB609AA/e/iOS.Simulator.Interpreter.Test.app/Contents/Info.plist
[16:14:49] dbug: Process PlistBuddy exited with 0
[16:14:49] info: Starting 'iOS.Simulator.Interpreter.Test' on MacCatalyst
[16:14:49] dbug: *** Executing 'iOS.Simulator.Interpreter.Test' on MacCatalyst ***
[16:14:49] dbug: 
[16:14:49] dbug: Running chmod +x /tmp/helix/working/BDE009DC/w/ADB609AA/e/iOS.Simulator.Interpreter.Test.app/Contents/MacOS/iOS.Simulator.Interpreter.Test
[16:14:49] dbug: Process chmod exited with 0
[16:14:49] dbug: 
[16:14:49] dbug: Running /System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Support/lsregister -f /tmp/helix/working/BDE009DC/w/ADB609AA/e/iOS.Simulator.Interpreter.Test.app
[16:14:49] dbug: Process lsregister exited with 0
[16:14:49] dbug: 
[16:14:49] dbug: Running open -W /tmp/helix/working/BDE009DC/w/ADB609AA/e/iOS.Simulator.Interpreter.Test.app
[16:14:55] dbug: Process open exited with 0
[16:14:55] info: App run ended, no abnormal exit code detected (0 assumed)
[16:14:55] fail: Application has finished with exit code 0 but 42 was expected
[16:14:55] dbug: Saving diagnostics data to '/tmp/helix/working/BDE009DC/w/ADB609AA/e/diagnostics.json'
XHarness exit code: 71 (GENERAL_FAILURE)
```

premun commented:
So the issue is that on all Apple platforms (iOS, tvOS, MacCatalyst), we have problems detecting apps exit code. The original (Xamarin) way of testing these platforms is always via the XHarness TestRunner that is included in the apps.

For all scenarios where we don't have the test runner (HelloiOS, functional tests), where we only fire up the app and wait for it to quit, XHarness is scanning MacOS syslogs for the event where the app terminated with non-zero code and parse the exit code from there. Unfortunately, this is not 100% reliable and sometimes we just don't see this line in the logs. We have no power over flushing of this log and there might be other reasons like MacOS cycling the log away when it reaches some size and starting a new one. 

We could implement some artifical waiting, though I it would be quite complicated. What I think would be better, since these tests only take few seconds, is to employ the DevWF retries and configure it to retry the run (can be on the same machine). I think this might help?

MaximLipnin commented:
> What I think would be better, since these tests only take few seconds, is to employ the DevWF retries and configure it to retry the run (can be on the same machine). I think this might help?

That makes sense, thanks. What I need to do is just to update https://github.com/dotnet/runtime/blob/main/eng/test-configuration.json with an appropriate rule for the functional tests, right?

premun commented:
@MaximLipnin something like that, yeah. But to be honest, I am not an expert in this area so not sure if this file goes where it needs and also if we can scope it to some platforms only for example. I might delegate you to the First Responder channel where people will be able to help you.

MaximLipnin commented:
@premun Thanks for helping! BTW, there would also be an option like adding `Console.WriteLine("exit-code=" + exitCode);` to a test and extending the xharness parsing logic, but Alexander called it an "ugly hack" 😄

premun commented:
Well, the whole Apple platform is one giant hack to be honest 😅 We have already a switch for a similar flow called `--signal-app-end` back from when we had some issues with detecting app shutting down, so we could theoretically extend it. However I am not a fan of this solution as then we are testing something a bit different. We can definitely bring back the discussion. However, if we manage to utilize the test retries here and it will work, I think it's a preferred way.

MaximLipnin committed:
Merge branch 'main' into move_maccatalyst_to_extra_platfroms_pipeline

MaximLipnin committed:
Configure the ios sumulator functional tests for re-trying in case of failure

MaximLipnin commented:
/azp run runtime-extra-platforms

steveisok commented:
Doesn't look like retries happened on the MacCatalyst failure.  @premun will we still see Attempt 1 and subsequent retries in the artifacts tab?

MaximLipnin commented:
Currently the test retry logic doesn't take into account our particular case 😢 

Apart from it,  another Alexander's suggestion: since xharness reads the whole /var/log/system.log, would it make sense to switch to calling the `log` command with a filter for the process ID?

premun commented:
Looking at the [docs (Test Retry Documentation.md)](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/Test%20Retry%20Documentation.md), it seems that the problem is that our work item doesn't return exit code 0 (point 4 in **How to get on board**).

I tried the filtering before but that won't solve the problem as it only leaves out lines of log which is already missing what we need. The non-filtered log we are scanning is actually very short (~10 lines) as there's not much of system logging during the few seconds the app runs. For example for the last build from this PR it's here:
- **MacCatalyst.system.log** - https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-64452-merge-4cfb40aa80854d4dab/iOS.Simulator.Interpreter.Test/1/MacCatalyst.system.log?sv=2019-07-07&se=2022-02-21T18%3A48%3A49Z&sr=c&sp=rl&sig=iReK24irme5NtAleRz2yYkXUVd3ERMa%2FSCTo4EYX42I%3D
(please note that in case of failures, we also attach the **system.log** grep that goes back a bit more so that's a different thing)

FWIW, the line we're looking for is this:
> Feb  1 13:50:17 dci-mac-build-183 com.apple.xpc.launchd[1] (net.dot.iOS.Simulator.Aot-Llvm.Test.164872[75355]): Service exited with abnormal code: 42

MaximLipnin commented:
/azp list

MaximLipnin committed:
Revert test retry configuration as it didn't work for the functional tests

MaximLipnin committed:
Merge branch 'main' into move_maccatalyst_to_extra_platfroms_pipeline

premun commented:
@MaximLipnin @akoeplinger we can add a new exit code for "Failed to detect exit code" and then have a retry in Helix SDK for those cases. I think that's the best we can do at the moment. XHarness CLI itself can warn the user about this.

premun commented:
I opened https://github.com/dotnet/xharness/issues/812

Should be piece of cake but will take couple of days to make changes in XHarness and Arcade and get here. I should be able to land this next week.

premun commented:
The XHarness and Arcade versions with the fix we need for this are `22104.1`

MaximLipnin committed:
Merge branch 'main' into move_maccatalyst_to_extra_platfroms_pipeline

MaximLipnin commented:
/azp run runtime-extra-platforms

MaximLipnin commented:
Now the xharness returns `82 (RETURN_CODE_NOT_SET)` in case of no exit code but the arcade dependency doesn't seem to be update to the right version yet.

premun commented:
Yeah, I think Arcade had hard time to get validated with the OSX outages. We can bump the Arcade in this PR to the right version if you want?

MaximLipnin commented:
Yes, nothing constraints us from doing this 😄  AFAIR the versions get bumped in eng/Version.Details.xml and eng/Versions.props. The version is 22104.1 and what is the right commit to specify?

premun commented:
You can navigate to your runtime folder locally and call this:
```
darc update-dependencies --source-repo dotnet/arcade --channel ".NET Eng - Validation" --id 125665
```

MaximLipnin commented:
Seems I have a problem with darc authentication

premun commented:
@MaximLipnin https://github.com/dotnet/arcade/blob/main/Documentation/Darc.md#setting-up-your-darc-client

MaximLipnin committed:
Merge branch 'main' into move_maccatalyst_to_extra_platfroms_pipeline

# Conflicts:
#	eng/pipelines/runtime-extra-platforms.yml

MaximLipnin committed:
Put MacCatalyst leg definitions to runtime-extra-platforms-other.yml

akoeplinger committed:
Bump arcade

MaximLipnin commented:
/azp run runtime-extra-platforms

MaximLipnin commented:
Now we can see retry attempts on MaCatalyst legs, but it didn't address the problem

premun commented:
@MaximLipnin can you try switching to the 10.15 queue?

MaximLipnin committed:
Merge branch 'main' into move_maccatalyst_to_extra_platfroms_pipeline

MaximLipnin commented:
> @MaximLipnin can you try switching to the 10.15 queue?

I guess, it's already done in https://github.com/dotnet/runtime/pull/65070

MaximLipnin commented:
/azp run runtime-extra-platforms

MaximLipnin commented:
MacCatalyst x64 legs seem to pass.
MacCatalyst arm64 failed with `Helix API does not contain an entry for OSX.1015.ARM64.Open` .

premun commented:
I fixed the wrong queue name in #65093.

The arm64 will have to go to 11.00 which might not pass as well as 10.15 does. Can we, for the time being, ignore that leg and move the rest? I will work on the exit codes and maybe we'll have a solution for it soon.

MaximLipnin committed:
Merge branch 'main' into move_maccatalyst_to_extra_platfroms_pipeline

MaximLipnin committed:
Leave MacCatalyst_arm64 in runtime-staging as it might not pass yet on OSX 11.00 queue

MaximLipnin commented:
> I fixed the wrong queue name in #65093.
> 
> The arm64 will have to go to 11.00 which might not pass as well as 10.15 does. Can we, for the time being, ignore that leg and move the rest? I will work on the exit codes and maybe we'll have a solution for it soon.

Updated to keep arm64 in runtime-staging for now

MaximLipnin commented:
/azp run runtime-extra-platforms

steveisok reviewed (approved):
LGTM

MaximLipnin commented:
Do we need to wait for https://github.com/dotnet/runtime/pull/65068 which is going to update the arcade dependency to the latest version?

premun commented:
@MaximLipnin there is a blocker and the dependency PR will take one more day to complete. However, we cannot merge this PR as it would break runtime due to some EOL errors. So we need to wait.

MaximLipnin committed:
Merge branch 'main' into move_maccatalyst_to_extra_platfroms_pipeline

# Conflicts:
#	eng/Version.Details.xml
#	eng/Versions.props
#	global.json

MaximLipnin commented:
/azp run runtime-extra-platforms

MaximLipnin committed:
Skip System.Reflection.Emit.Tests.ILGeneratorEmit4.TestDynamicMethodEmitCalliNonBlittable test failing on MacCatalyst x64

MaximLipnin commented:
/azp run runtime-extra-platforms

MaximLipnin committed:
Revert skipping System.Reflection.Emit.Tests.ILGeneratorEmit4.TestDynamicMethodEmitCalliNonBlittable test on MacCatalyst x64 as the respective issue is addressed

MaximLipnin committed:
Merge branch 'main' into move_maccatalyst_to_extra_platfroms_pipeline

MaximLipnin commented:
/azp run runtime-extra-platforms

MaximLipnin commented:
Failures are unrelated, merging
